### PR TITLE
Pytest migration

### PR DESCRIPTION
**Pull request by zdutta**
Migrating testing suite from nose to pytest. Split into 4 commits: functional, integration, unit pt1 and unit pt2

### CODE REVIEW

**Review by nateprewitt — CHANGES_REQUESTED** [review]
Only about 1/5 of the way through but wanted to leave this now to cut down on iterations.

File: tests/functional/csm/test_monitoring.py [hunk 1]
```diff
@@ -32,6 +33,7 @@
 
 CASES_FILE = os.path.join(os.path.dirname(__file__), 'cases.json')
 DATA_DIR = os.path.join(os.path.dirname(__file__), 'data/')
+TEST_CASES = []
```
[the remark this answers is not in the document]
Comment (joguSD) [reply tests/functional/csm/test_monitoring.py:36]: I think this just forgot to be deleted.

File: tests/functional/retries/test_bucket.py [hunk 2]
```diff
@@ -98,12 +98,12 @@ def test_stress_test_token_bucket(self):
...
+        assert self.caught_exceptions == []
         distribution = self.acquisitions_by_thread.values()
         mean = sum(distribution) / float(len(distribution))
         # We can't really rely on any guarantees about evenly distributing
         # thread acquisition(), e.g. must be with a 2 stddev range, but we
         # can sanity check that our implementation isn't drastically
         # starving a thread.  So we'll arbitrarily say that a thread
         # can't have less than 20% of the mean allocations per thread.
-        self.assertTrue(not any(x < (0.2 * mean) for x in distribution))
+        assert any(x < (0.2 * mean) for x in distribution) is False
```
Comment (joguSD) [inline tests/functional/retries/test_bucket.py:109]: Any reason to to not use `assert not any`?

File: tests/functional/test_apigateway.py [hunk 3]
```diff
@@ -33,6 +33,5 @@ def test_get_export(self):
         with self.http_stubber:
             self.client.get_export(**params)
             request = self.http_stubber.requests[0]
-            self.assertEqual(request.method, 'GET')
-            self.assertEqual(request.headers.get('Accept'),
-                             b'application/yaml')
+            assert request.method == 'GET'
+            assert request.headers.get('Accept') == b'application/yaml'
```
Comment (joguSD) [inline tests/functional/test_apigateway.py:37]: Kind of strange that the newline got removed.

File: tests/functional/test_cloudsearchdomain.py [hunk 4]
```diff
@@ -26,7 +26,7 @@ def test_search(self):
         with self.http_stubber:
             self.client.search(query='foo')
             request = self.http_stubber.requests[0]
-            self.assertIn('q=foo', request.body)
-            self.assertEqual(request.method, 'POST')
+            assert 'q=foo' in request.body
+            assert request.method == 'POST'
             content_type = b'application/x-www-form-urlencoded'
-            self.assertEqual(request.headers.get('Content-Type'), content_type)
+            assert request.headers.get('Content-Type') == content_type
```
Comment (joguSD) [inline tests/functional/test_cloudsearchdomain.py:32]: Another removal of the new line.

File: tests/functional/test_credentials.py [hunk 5]
```diff
@@ -68,13 +69,13 @@ def assert_consistent_credentials_seen(self, creds, func):
             # from the first refresh ('1'), the secret key from
             # the second refresh ('2'), and the token from the
             # first refresh ('1').
-            self.assertTrue(creds[0] == creds[1] == creds[2], creds)
+            assert creds[0] == creds[1] == creds[2]
```
Comment (joguSD) [inline tests/functional/test_credentials.py:72]: Is there a particular you ran into that made you remove the failure messages when porting to `assert`?

File: tests/functional/test_credentials.py [hunk 6]
```diff
@@ -319,7 +318,7 @@ def test_environment_credential_source(self):
         self.assert_creds_equal(actual_creds, expected_creds)
 
         stubber.assert_no_pending_responses()
-        self.assertEqual(self.env_provider.load.call_count, 1)
+        self.env_provider.load.call_count == 1
```
Comment (joguSD) [inline tests/functional/test_credentials.py:321]: Looks like we forgot to `assert` here.

File: tests/integration/test_emr.py [hunk 7]
```diff
@@ -11,22 +11,23 @@
...
 from tests import unittest
+import pytest
 
 import botocore.session
 from botocore.paginate import PageIterator
 from botocore.exceptions import OperationNotPageableError
 
 
-def test_emr_endpoints_work_with_py26():
+@pytest.mark.parametrize('region',['us-east-1', 'us-west-2', 'ap-northeast-1',
```
Comment (joguSD) [inline tests/integration/test_emr.py:21]: Might be easier to just pull this list out as a `EMR_REGIONS` variable at the module level or something.

File: tests/integration/test_emr.py [hunk 8]
```diff
@@ -47,18 +48,18 @@ def test_can_access_pagination_configs(self):
         # Using an operation that we know will paginate.
         paginator = self.client.get_paginator('list_clusters')
         page_iterator = paginator.paginate()
-        self.assertIsInstance(page_iterator, PageIterator)
+        assert isinstance(page_iterator ,PageIterator)
```
Comment (joguSD) [inline tests/integration/test_emr.py:51]: Nit: looks like the whitespace around the comma got flipped.

**Comment by joguSD** [comment]
Overall the changes seem reasonable, looks like there's some failures though? Maybe related to a parameterize migration? 
Also looks like the CI is broken on 2.7, though we're still using travis on this branch. Let's get this passing on Python 3 and then we'll merge and rebase the migration branch to include the migration to github actions and we'll address fixing the 2.7 failure the

**Comment by joguSD** [comment]
Overall the changes seem reasonable, looks like there's some failures though? Maybe related to a parameterize migration? 
Also looks like the CI is broken on 2.7, though we're still using travis on this branch. Let's get this passing on Python 3 and then we'll merge and rebase the migration branch to include the migration to github actions and we'll address fixing the 2.7 failures then.

**Comment by joguSD** [comment]
This looks good to me, I'm not too concerned about travis failing right now -- let's get CI fully functional again once we rebase on top of develop and make sure we're working under github actions.

**Review by joguSD — APPROVED** [review]

**Pull request merged**